complete update for 0.7 and lots of test fixes

ExpandingMan opened the pull request:
Hello all, I'm not done with this yet but I wanted to put the PR up now so people know that it's being worked on and I have made significant progress.

Apart from the fixes for 0.7 I have noticed that the tests are quite a mess right now.  It is very difficult to get 0.7 up and working without clean tests, so I am prioritizing getting the tests nice and spiffy.

I am fixing a huge number of deprecation warnings (that originate in DataFrames.jl) in the tests.  I would like to suggest that if we want to keep tests of deprecated functionality, these should be moved to a dedicated test module because otherwise it makes cleaning up tests that are actually problematic rather difficult.  This can also cover up real issues.  For instance, does `getindex` require a `makeunique` keyword?  Suppose you do `df[:, [1,1,2]]` as in the tests?  I'll post an issue about this (if there isn't one already).

I may not get to finish this until the end of the week, but I will post updates soon.

Wow, what timing, looks like there will be significant overlap between here and #1363.  Perhaps I'll go back and see if I can branch off of that.

nalimilan commented on src/other/index.jl:
```diff
@@ -158,7 +158,7 @@ function Base.getindex(x::AbstractIndex, idx::AbstractVector)
     end
     length(idxs) == 0 && return Int[] # special case of empty idxs
     if idxs[1] isa Real
-        if !all(v -> v isa Integer && !(v isa Bool), idxs)
```
Why change this?

ExpandingMan replied:
because it was breaking when a `Symbol` was passed in

ExpandingMan replied:
Nothing, it was breaking in 0.6 also.  Since it only produced a warning nobody seems to have noticed.

nalimilan replied:
@bkamins Does that look OK to you?

bkamins replied:
It was breaking because we do not allow mixing indexing by number and symbol in the same index. It has to be either all numbers or all symbols. We have to exclude `Bool` because we do not allow indexing by `Bool` in this case and `Bool <: Integer`.

Now conversion to `Int` will fail on `Symbol` anyway in line 169.

ExpandingMan replied:
oh, I think I was *very* confused about what the intent was there.  will change.

nalimilan commented on src/other/utils.jl:
```diff
@@ -46,7 +46,9 @@ function makeidentifier(s::AbstractString)
 
-    return String(take!(res))
+    # return String(take!(res))
+    # TODO this is to fix a bug in Compat for 0.6
```
Can you give more details? Is there an issue open in Compat.jl?

ExpandingMan replied:
see [this issue](https://github.com/JuliaLang/Compat.jl/issues/502).  I'm hoping it will be resolved one way or another before I'm even done with this PR...

nalimilan commented on src/DataFrames.jl:
```diff
@@ -9,14 +9,23 @@ module DataFrames
+        esc(:(isdefined($sym)))
+    end
+    using Compat.IOBuffer
+    const normalize = normalize_string  # requires that we don't use LinearAlgebra normalize
```
I'd rather wrap this inside `module Unicode`, and avoid `using Unicode: normalize` in the other branch, so that you write `Unicode.normalize` everywhere to avoid the ambiguity with `LinearAlgebra.normalize`.

ExpandingMan replied:
will do...

ExpandingMan replied:
Julia is not letting me define a module in a conditional statement.  Any other ideas for dealing with this?  I really want to try hard to put only valid 0.7 code in the main body.

nalimilan commented on src/DataFrames.jl:
```diff
@@ -9,14 +9,23 @@ module DataFrames
     const kwpairs = pairs
 else
     kwpairs(x::AbstractArray) = (first(v) => last(v) for v in x)
+    macro isdefined(sym)
```
Would be nice to contribute this to Compat if you have the time.

ararslan replied:
I did that a while back but it was rejected

ExpandingMan replied:
yeah, as far as I know `@isdefined` has been around for quite a while, so I think there's some reason it's not in `Compat`, though I have no idea what that is.  @ararslan , do you think you could find that issue or thread?

ararslan replied:
https://github.com/JuliaLang/Compat.jl/pull/402

ExpandingMan replied:
ok, can we perhaps keep it in DataFrames though?  I'm trying very hard to put *only* 0.7 valid code into the main body so that we don't really have to do anything once we no longer support 0.6 (which considering the nature of 0.7 may not be very long)

nalimilan replied:
OK, let's discuss this in Compat. But here I think you can just do `isdefined(Base, :unique!)`, the fact that `Base` was omitted is most likely an oversight.

ExpandingMan commented:
Please don't hesitate to merge #1363 on account of this PR.  I plan to incorporate those changes as well...

ExpandingMan commented on src/DataFrames.jl:
```diff
@@ -99,6 +108,6 @@ include("dataframerow/show.jl")
 include("dataframe/sort.jl")
 
-include("deprecated.jl")
+# include("deprecated.jl")
```
commenting this out was done for testing purposes... will be restored, please ignore

ExpandingMan commented:
Can I remove code for pre 0.6?  see [here](https://github.com/JuliaData/DataFrames.jl/blob/master/src/other/utils.jl#L9)

ExpandingMan commented:
Ok, this has been way worse than I was expecting, but I'm slowly getting there.

CategoricalArrays is in pretty bad shape for 0.7, looks like there will be a ton of warnings and even a couple of errors that may have to wait for that.

The `IOBuffer` issue fix is being pushed to compat, see [here](https://github.com/JuliaLang/Compat.jl/pull/504).

ExpandingMan commented:
Ok!  I have cleaned everything up the best I think is currently possible.  Some comments:
- CategoricalArrays and StatsBase are both still badly screwed up on 0.7.  There are a number of 0.7 test errors and deprecation warnings resulting from CategoricalArrays.  If I have time I'm going to clean up these packages also.
- All 0.6 tests now pass without any errors or unexpected warnings.  I have moved all tests of deprecated functionality to the new test file "deprecated.jl".  I very much would like to keep all tests of deprecated functionality separate, as it becomes very difficult to parse the test results on 0.7 when there are lots of deliberate deprecation warnings thrown in.
- The `IOBuffer` bug on Compat has been fixed so I have removed my hack.

Hopefully we can avoid master having DataFrames related test errors in the future? :wink:

Tagging @nalimilan , please let me know any changes you'd like me to make so we can get this merged.

ExpandingMan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -781,6 +782,11 @@ Base.hcat(df1::AbstractDataFrame, df2::AbstractDataFrame, dfn::AbstractDataFrame
 
+
+# TODO this function to be removed after CategoricalArrays update
+_promote_col_type(::Type{T}, n::Integer) where T = T(uninitialized, n)
```
To make the semantics consistent, they should support the `T(uninitialized, n)` constructor.  They don't necessarily *have* to do that, as far as I know Julia isn't necessarily trying to enforce this as a standard.  Of course, I think they should support it.  It's a design decision that will have to be made at some point, but for now, here's a work-around.

nalimilan replied:
OK, I've prepared https://github.com/JuliaData/CategoricalArrays.jl/pull/130.

ExpandingMan replied:
On second thought I'm not even sure defining this constructor for `CategoricalArray` would be consistent.  See my response in your PR.

ExpandingMan commented on src/dataframe/dataframe.jl:
```diff
@@ -121,6 +121,8 @@ function DataFrame(pairs::Pair{Symbol,<:Any}...; makeunique::Bool=false)::DataFr
     DataFrame(columns, Index(colnames, makeunique=makeunique))
 end
 
+DataFrame(dict::Dict{Symbol,<:Any}) = DataFrame((x for x ∈ dict)...)
```
Oh, this was really interesting I forgot to mention it.  0.7 throws a deprecation warning without this explicit constructor because the "automatic conversion" constructors are deprecated, but for the life of me I have no idea what the `Dict` constructor actually wound up calling before.  My first instinct was to do away with the `Pair` constructor and have it feed into a `Dict` constructor, but I had to do it the other way around because of `makeunique`.

nalimilan replied:
OK. Then `Base.convert(::Type{DataFrame}, d::AbstractDict)` was called automatically on 0.6. On 0.7, the recommendation added in https://github.com/JuliaLang/julia/pull/23273 is to define constructors, and have `convert` methods call constructors explicitly (so the inverse of what we do now, mostly).

nalimilan replied:
(Note that splatting should generally be avoided when the number of elements can be large, as is the case here: it's much more efficient to pass the collection and loop over it inside the function.)

ExpandingMan commented on src/abstractdataframe/io.jl:
```diff
@@ -139,10 +139,11 @@ function latex_char_escape(char::AbstractString)
     end
 end
+# in 0.7 this function will only ever get passed Char, in 0.6 always gets SubString
+latex_char_escape(char::Char) = latex_char_escape(string(char))
```
Agreed.  0.6 *only* works with `SubString` and 0.7 *only* works with `Char`, so I'll have to think about the best way of doing this...

nalimilan replied:
The easiest solution is to have two completely separate blocks, one for each version, and duplicate the code as needed.

ExpandingMan commented on test/utils.jl:
```diff
@@ -10,9 +16,10 @@ module TestUtils
+    # TODO uncomment this line after deprecation period when makeunique=false throws error
     # @test_throws ArgumentError DataFrames.make_unique([:x, :x, :x_1, :x2], makeunique=false)
-    @test DataFrames.make_unique([:x, :x_1, :x2], makeunique=false) == [:x, :x_1, :x2]
+    # TODO below is commented out for now for warning sanity
```
lol, I don't even remember... will change

ExpandingMan commented on test/join.jl:
```diff
@@ -71,63 +72,56 @@ module TestJoin
     # Cross joins don't take keys
     @test_throws ArgumentError join(df1, df2, on = :A, kind = :cross)
 
-    @testset "Test empty inputs 1" begin
```
Was getting function overwrite warnings for `simple_df`

ExpandingMan replied:
I'm confused, we *are* overwriting the method, hence the warning.  I'm not sure why there'd be any a priori preference on these function names, was it meant to be extending something?  Do you want me to revert to the warnings?

nalimilan replied:
Actually I've not been able to reproduce the warning, so it's hard to understand what it means. On 0.7 `simple_df` isn't defined outside of `@testset` blocks, so everything should be fine.

Anyway we should definitely not remove the `@testset` blocks.

ExpandingMan replied:
Woah.... terribly sorry, legitimately no idea what happened here.  I may have screwed it up with abuse of vim editing tools, or perhaps github did something strange when I merged one of the other commits?  I'm very confused.  Sorry, restoring now, definitely agree that we should not remove `@testset` blocks!

ExpandingMan commented on test/index.jl:
```diff
@@ -13,11 +14,11 @@ inds = Any[1,
            [big(1)],
            big(1):big(1),
            [:A],
-           Union{Bool, Missing}[true, false],
```
throws deprecation warning for `Union{T,Missing}` vectors since indexing by `missing` is deprecated.  note that this is now tested in "deprecated.jl".

bkamins replied:
On my machine it does not throw deprecation warning. The vector does not contain any missing and `getindex` for vector of `Union{Bool, Missing}` tests for it. And this specific case is not tested in "deprecated.jl".

ExpandingMan replied:
Yes this should not have been removed, not sure why I did, will restore.  I'm actually pretty dubious about whether it makes sense to accept vectors of this type, but that's a separate issue so I'll file one.

bkamins replied:
Agreed 😄. When I was making a clean up of `getindex` I also wondered why we accept such unions, but we did so I left it (probably this is due to the fact that in some cases the indexing vector might be dynamically generated and some methods "by default" can create such unions without checking if they are needed).

nalimilan replied:
Yes, I think it makes sense to ignore the type of the array and look only at the actual type of the elements. Else it can be cumbersome given that Julia is a dynamic language. That's how indexing arrays works.

bkamins replied:
The only exception to keep in mind is the fact that `Bool <: Integer` and we want to handle `Bool` in a different way than `Integer`. It is now covered correctly by `getindex` (i.e. array consisting only of bool values uses boolean indexing only if its type is `AbstractVector{Bool}` or `AbstractVector{Union{Missing,Bool}}` and otherwise it is an error) but if someone changes this code it has to be kept in mind.

nalimilan replied:
The only way to ensure this doesn't regress is to add tests...

bkamins replied:
Agreed. @ExpandingMan could you please make sure that we have the following tests?
Otherwise I can add it later, but maybe it is better to fix it in one shot.

Tests that should pass after deprecation:
```
[true, false]
Union{Missing,Bool}[true, false]
[false:false]
false:true
```
(the last one is a corner case but is OK in this particular case)

Tests that should fail:
```
[true]
[true, false, true]
[true, 1]
Any[true, true]
Integer[true, false]
[missing, true]
Union{Missing, Bool}[missing, missing]
```

nalimilan replied:
`[true, 1]` should be `Any[true, 1]` or it's equivalent to `[1, 1]`.

ExpandingMan commented on test/cat.jl:
```diff
@@ -51,23 +57,30 @@ module TestCat
     @testset "hcat ::AbstractVectors" begin
         df = DataFrame()
-        DataFrames.hcat!(df, CategoricalVector{Union{Int, Missing}}(1:10))
+        global df2 = DataFrame(x1=CategoricalVector{Union{Int,Missing}}(1:10))
```
Is global causing a problem?  I would prefer to just rename everything instead of living with the warning if that's what it takes.  The warnings are more than just annoying, they make it very difficult to spot real problems, especially in 0.7.0.

The original `hcat` functions were causing deprecation warnings.  Might be fixed now with alternate naming, let me try changing it back.

nalimilan replied:
Julia 0.7 will go away in a few months, but the code base is going to stay for years. So I'd rather not add incorrect statements just to silence warnings. Clearly `df2` isn't global here. It's fine to use this temporarily to help finding real warnings, but better not include it in the final PR.

ExpandingMan replied:
Ok, I'm getting rid of all of the `global`s now.  I'm renaming all of the variables so that they don't produce warnings.

ExpandingMan commented on test/data.jl:
```diff
@@ -1,14 +1,19 @@
 module TestData
     using Compat, Compat.Test, DataFrames
-    importall Base # so that we get warnings for conflicts
+    # TODO this already generates a huge wall of warnings
```
Ok, I'm not really sure what the best solution is here.  Calling `importall Base` for the purpose of generating warnings when it already generates a ton of warnings even when working properly seems crazy.  Is there some aspect of this we can move to the new `deprecated.jl`?

nalimilan replied:
What happens if you just remove it?

ExpandingMan commented:
Hello all, I'd like to make a general plea that warnings be taken more seriously in the future.

Getting this going on 0.7 was surprisingly painful and that was much exacerbated by the many warnings already appearing in the tests under normal circumstances.  Not only to these make updates more difficult, but they can hide real issues, a couple of examples of which can be seen in this PR.  This also makes it a lot harder to notice legitimate warnings from Base Julia concerning needed changes and warnings that might be inappropriately given to users.  Lastly, the warnings make it more difficult for people to contribute to the package, since it can be hard to figure out what then intent of some of the tests and warnings are (as it sometimes was for me).

Note that I have moved all tests of deprecated functionality which deliberately produce warnings to a separate `deprecated.jl` test file, which should make it easier to filter out some of the intended warnings.

bkamins commented on test/dataframe.jl:
```diff
@@ -51,17 +51,7 @@ module TestDataFrame
-    z = vcat(v, x)
-
-    # TODO: uncomment the line below after deprecation
-    # @test_throws ArgumentError z[:, [1, 1, 2]]
```
Why this line (61 and 62) is removed? I think we should test that it throws an error after deprecation.

ExpandingMan replied:
Yes, but currently this is not fully deprecated behavior so it does not yet throw an error.  When automatic generation of column names from indexing is fully deprecated, this line should be un-commented.  Note that we test the current behavior in `deprecation.jl`.

bkamins commented:
@ExpandingMan Huge work. Thank you. I have left my comments inline.

nalimilan commented:
I agree that warnings are problematic in general. I don't like warnings because they indicate something is wrong, yet no failure happens. Ideally we would turn deprecations from DataFrames into errors (but not those from other packages), like Julia does, but I don't think that's possible currently, and tests for deprecated features would not be possible.

ExpandingMan commented on test/join.jl:
```diff
@@ -37,17 +38,17 @@ module TestJoin
-    @test join(nameid, jobid, on = :ID, kind = :right) == right[on]
-    @test join(nameid, jobid, on = :ID, kind = :semi) == semi[on]
-    @test join(nameid, jobid, on = :ID, kind = :anti) == anti[on]
+    on_ = [:ID]
```
All such renames are in order to avoid extraneous warnings.  Do you want me to revert to warnings?

nalimilan replied:
Yes, as I said I prefer keeping the code clean and wait until the (spurious) warnings go away than living with weird variable names forever.

bkamins commented:
@ExpandingMan Regarding your question about `df[:, [1,1,2]]` I think that `getindex` does not need `makeunique` keyword argument. Such duplicate columns are aliases and I do not see much use for such indexing, but I might not see something so it is not a strong no-go.

ExpandingMan commented:
Yeah, my confusion has cleared up a bit.  It seems that `df[:, [1,1,2]]` is deprecated, currently it throws a warning, and it will throw an error in the future.  There is already a test of the error that should be uncommented when the full deprecation is enacted.

bkamins commented:
Exactly! Thanks

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -121,6 +121,8 @@ function DataFrame(pairs::Pair{Symbol,<:Any}...; makeunique::Bool=false)::DataFr
     DataFrame(columns, Index(colnames, makeunique=makeunique))
 end
 
+DataFrame(dict::AbstractDict) = convert(DataFrame, dict)
```
You appear to have missed [my previous comment](https://github.com/JuliaData/DataFrames.jl/pull/1364#discussion_r171516203).

ExpandingMan replied:
? Wasn't it regarding splatting?  This is no longer splatting.  Can you explain what you'd like me to change here?

nalimilan replied:
I mean this:
> On 0.7, the recommendation added in JuliaLang/julia#23273 is to define constructors, and have convert methods call constructors explicitly (so the inverse of what we do now, mostly).

ExpandingMan replied:
Oh, sorry I had missed that.  Will change now.

nalimilan commented on src/abstractdataframe/io.jl:
```diff
@@ -130,19 +130,30 @@ end
-        return "\\textasciitilde{}"
-    else
-        return string("\\", char)
+if VERSION ≥ v"0.7.0-"
```
That's `0.7.0-DEV.4059`.

ExpandingMan commented on src/DataFrames.jl:
```diff
@@ -9,16 +9,23 @@ module DataFrames
 else
     kwpairs(x::AbstractArray) = (first(v) => last(v) for v in x)
+    using Compat.IOBuffer
+    import Base.LinAlg: normalize
```
There was a Compat update so I was really hoping I could do this properly, but it's still really screwed up.  When you do `using Compat.Unicode: normalize` in 0.6 it doesn't import the `normalize_str` methods.  Anyway, this can't be removed, because it throws an error telling you that you must explicitly import `LinAlg.normalize`.  Sorry I've been hesitant to make PR's to Compat, but there are so many gotchas and existing bugs that doing that is rather difficult.

This should *not* break if Compat gets fixed, especially because `Unicode` doesn't even export `normalize`.

nalimilan replied:
OK. Doesn't it work if you do `using Compat.Unicode` and then `Unicode.normalize` for every use?

nalimilan replied:
BTW, you should be able to use it soon (don't forget to update REQUIRE): https://github.com/JuliaLang/METADATA.jl/pull/13615

ExpandingMan replied:
awesome, thanks, I will wait until METADATA gets updated and try it out.

ExpandingMan commented on test/data.jl:
```diff
@@ -1,12 +1,13 @@
-    using Compat, Compat.Test, DataFrames
-    importall Base # so that we get warnings for conflicts
+    using Compat, Compat.Test, DataFrames, Compat.Random
+    # TODO: need to find a better way of accomplishing below
```
You want me to retain the `importall` or remove it?

nalimilan replied:
I'd say remove it, anyway the comment probably referred to ambiguity warnings, which are no longer printed.

nalimilan commented on test/dataframe.jl:
```diff
@@ -51,17 +51,7 @@ module TestDataFrame
-    #
-
     x = DataFrame(a = [1, 2, 3], b = [4, 5, 6])
-    v = DataFrame(a = [5, 6, 7], b = [8, 9, 10])
```
Still need to do something about this: https://github.com/JuliaData/DataFrames.jl/pull/1364/files#r171626211

ExpandingMan replied:
I did, it's in `deprecated.jl` now.  This dataframe `v` is removed here because it is no longer used in this file (appears again in `deprecated.jl`).

nalimilan commented on test/cat.jl:
```diff
@@ -51,23 +51,25 @@ module TestCat
         @test df[3] == collect(1:10)
 
         df = DataFrame()
-        df2 = hcat(CategoricalVector{Union{Int, Missing}}(1:10), df)
```
Can you revert all the renames in this file and the following files?

ExpandingMan replied:
Yes, don't worry, I'm not ignoring you, I was busy...

nalimilan replied:
No worries. I leave new comments because GitHub hides the old ones and it's hard to keep track.

ExpandingMan replied:
Yeah, I also never know if you are aware that I made changes, though by now it's seeming that you must get a notification whenever I commit...

nalimilan commented on test/grouping.jl:
```diff
@@ -171,9 +171,9 @@ module TestGrouping
-        df = DataFrame(Key1 = ["A", missing, "B", "B", "A"],
-                       Key2 = CategoricalArray(["B", "A", "A", missing, "A"]),
-                       Value = 1:5)
+        df= DataFrame(Key1 = ["A", missing, "B", "B", "A"],
```
Wrong indentation.

nalimilan commented on test/show.jl:
```diff
@@ -284,11 +284,11 @@ module TestShow
     ├───────┼──────┼──────────────────────────────────┼─────────┼─────────────────┤
     │ 1     │ Fish │ String                           │ 0       │ Suzy  …  Amir   │
-    │ 2     │ Mass │ Union{Float64, Missings.Missing} │ 1       │ 1.5  …  missing │"""
+    │ 2     │ Mass │ Union{Float64, $Missing} │ 1       │ 1.5  …  missing │"""
```
I've just realized this probably won't work on 0.7, where the `Missing` prefix won't be present, as the other columns will be less wide than in the string. So I guess we'll need two blocks with a `VERSION` check. There's another case below.

ExpandingMan replied:
good catch!

ExpandingMan commented:
Thanks for pushing that last commit.  Sorry, as this wraps up I'm starting to lose track of things.

ExpandingMan commented:
Yay!  In the future if I ever do something this big I'll probably split it into multiple PR's.  I suspect that might save us all a bit of sanity.

nalimilan commented:
Hmm, I wonder why some lines are not covered by tests anymore. For example, compare:
https://coveralls.io/builds/15784016/source?filename=src%2Fdataframe%2Fdataframe.jl#L403
with the last commit on master:
https://coveralls.io/builds/15401106/source?filename=src%2Fdataframe%2Fdataframe.jl#L403

ExpandingMan commented:
That was added for a reason, though now I don't remember exactly what that was.  I expect this to break something.

nalimilan commented:
Yeah, there's a deprecation warning about it, but let's handle this in the next round of fixes. Else we need more tests, which is going to hold off the PR.

ExpandingMan commented:
Oh! I remember what it fixes.

Without it you can't do e.g. `df[1:end, 1]`.

nalimilan commented:
Let's handle this in another PR, anyway we're not there yet!

ExpandingMan commented:
Acknowledged, opened an issue so we don't forget.
